File: hardware/code/miscellaneousTests/Calc_staple_test.cpp
/*
 * Copyright (c) 2011,2012,2015,2016 Christopher Pinke
 * Copyright (c) 2013 Matthias Bach
 * Copyright (c) 2016 Francesca Cuteri
 * Copyright (c) 2018 Alessandro Sciarra
 *
 * This file is part of CL2QCD.
 *
 * CL2QCD is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * CL2QCD is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with CL2QCD. If not, see <http://www.gnu.org/licenses/>.
 */

// use the boost test framework
#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE staple_test
#include "testCode.hpp"

#include <boost/test/unit_test.hpp>

const ReferenceValues calculateReferenceValue_staple(const LatticeExtents lE)
{
    return ReferenceValues{-11.30184821830432 * lE.getLatticeVolume()};
}

struct StapleTestCode : public TestCode {
    StapleTestCode(const hardware::code::OpenClKernelParametersInterface& kP, hardware::Device* device)
        : TestCode(kP, device)
    {
        testKernel = createKernel("staple_test") << get_device()->getGaugefieldCode()->get_sources()
                                                 << "../hardware/code/miscellaneousTests/staple_test.cl";
    }

    virtual void runTestKernel(const hardware::buffers::SU3* gf, const hardware::buffers::Plain<hmc_float>* out,
                               const int gs, const int ls) override
    {
        err = clSetKernelArg(testKernel, 0, sizeof(cl_mem), gf->get_cl_buffer());
        BOOST_REQUIRE_EQUAL(CL_SUCCESS, err);
        err = clSetKernelArg(testKernel, 1, sizeof(cl_mem), out->get_cl_buffer());
        BOOST_REQUIRE_EQUAL(CL_SUCCESS, err);

        get_device()->enqueue_kernel(testKernel, gs, ls);
    }
};

struct StapleTester : public OtherKernelTester {
    StapleTester(const ParameterCollection pC, const GaugefieldTestParameters tP)
        : OtherKernelTester("StapleTest", pC, tP, calculateReferenceValue_staple(tP.latticeExtents))
    {
        testCode = new StapleTestCode(pC.kernelParameters, device);
        testCode->runTestKernel(OtherKernelTester::gaugefieldBuffer, out, gs, ls);
    }
};

BOOST_AUTO_TEST_CASE(STAPLE_TEST)
{
    GaugefieldTestParameters parametersForThisTest{LatticeExtents{4, 4}, GaugefieldFillType::nonTrivial};
    hardware::HardwareParametersMockup hardwareParameters(parametersForThisTest.ns, parametersForThisTest.nt);
    hardware::code::OpenClKernelParametersMockup kernelParameters(parametersForThisTest.ns, parametersForThisTest.nt);
    ParameterCollection parameterCollection(hardwareParameters, kernelParameters);
    StapleTester tester(parameterCollection, parametersForThisTest);
}

File: tools/mathematica/referenceValuesForTests/hardware/code/miscellaneousTests/calcStapleTest.m
(* ::Package:: *)

(*
 * Copyright 2012,2013 Lars Zeidlewicz,Christopher Pinke,
 * Matthias Bach,Christian Sch\[ADoubleDot]fer,Stefano Lottini,Alessandro Sciarra
 *
 * This file is part of CL2QCD.
 *
 * CL2QCD is free software:you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation,either version 3 of the License,or
 * (at your option) any later version.
 *
 * CL2QCD is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY;without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with CL2QCD.If not,see<http://www.gnu.org/licenses/>.
*)


SetOptions[SelectedNotebook[], PrintPrecision -> 16]

(*Get the needed packages for these tests*)
Get[FileNameJoin[{ParentDirectory[ParentDirectory[ParentDirectory[ParentDirectory[NotebookDirectory[]]]]], "packages/threeBythreeMatrices.m"}], Path -> {NotebookDirectory[]}]


(*Staple_sum*)
stSum[u_] :=
  3*(u.ConjugateTranspose[u].ConjugateTranspose[u] +
        ConjugateTranspose[u].ConjugateTranspose[u].u)
(*Calc_staple tests*)
(Re[Total[stSum[cold3x3mat], {1, 2}]] + Im[Total[stSum[cold3x3mat], {1, 2}]]) * 4
(Re[Total[stSum[nonTrivial3x3mat], {1, 2}]] + Im[Total[stSum[nonTrivial3x3mat], {1, 2}]]) * 4
